### PR TITLE
Add support for network interface power management

### PR DESCRIPTION
**Pull request by fhunleth**
This adds the VintageNet.PowerManager behaviour and allows projects to
register implementations with VintageNet. VintageNet will call the
implementation to power on and off the hardware as needed. For example,
if a device isn't working, VintageNet can power it off and back on.

### CODE REVIEW

**Comment by fhunleth** [comment]
@ConnorRigby @mattludwigs While there's one more thing that I'd like to validate with power management, I think this is ready for review. I think that the easiest way to review is to look at the example in the cookbook first and then go through the docs.

I think the main issue with power management is when does it decide to cycle power on a device. Right now, the logic is to turn on a device when there's a config that uses it, turn it off if that config goes away, and cycle power if the config changes or can't be applied. All power cycling is moderated by timing parameters in the PowerManager implementation so this isn't as frequent as it sounds. 

In order to recover from issues, calls to cycle power need to be added in more places. I have not done that since I haven't thought it through all the way. I'm thinking that I may put that in a separate PR since this one is already big and complicated. To get you a feel for possible ways, it could be things like the network interface has not been able to connect to an AP or cell tower for 5 minutes - therefore, power cycle the network adapter and try again. What's making me hesitate on that is that I've seen LTE modems report that they're connected but not pass data. A reboot of the modem fixes that (a disconnect/reconnect via AT commands does too, but that doesn't feel like as complete a fix as a reboot).

File: docs/cookbook.md [hunk 1]
```diff
@@ -301,3 +301,56 @@ VintageNet.deconfigure("wlan0", persist: false)
...
+  @behaviour VintageNet.PowerManager
+
+  @power_on_hold_time 5 * 60000
+  @min_powered_off_time 5000
+
+  defstruct reset_n: nil
+
+  @impl VintageNet.PowerManager
+  def init(_args) do
+    {:ok, reset_n} = Circuits.GPIO.open(@reset_n_gpio, :output)
```
Comment (mattludwigs) [inline docs/cookbook.md:324]: I might be missing something but I don't see `@reset_n_gpio` defined?

As the value of the `@reset_n_gpio` will be user-defined and specific to the hardware I can see a reason for not setting it in this example. When was reading this I was a little thrown off and started looking for any other macros that would inject this module attribute. However, that might just be me.

Comment (fhunleth) [reply docs/cookbook.md:324]: Ugh, that's a little depressing that macro injection of module attributes is something that people do enough to make you think that it could happen. I'll add it to avoid any concern about magic.

**Comment by fhunleth** [comment]
FYI - I've been experiment with this and there are a few things that are counterintuitive. For example, say you request a reset since things looks like they're not working and the reset takes time to be issued due to mandatory hold times. If the network starts working, then the reset still happens. It seems like what's needed is a `cancel_reset` that can be called when you know everything is good.

This leads me to believe that a better strategy for this whole thing is to make it more like a watch dog. I.e., code needs to actively report that hardware is working or it will be reset. For example, for LTE modems, regular polling using AT commands for connectivity seems like it could do this. I can imagine ways of doing it for WiFi and wired Ethernet should that be necessary. 

What that means is that the `PMControl` interface needs to be changed so that `reset` is removed and `power_on` is added. VintageNet calls `power_on` whenever a non-null config uses the interface. Then we need to find good places where it's known that the interface is working and put `pet` calls there. 

In summary, I feel like I need to invert the problem so that network interfaces are assumed broken unless proven otherwise. The current state of identifying places where the interface is known broken is pretty hard to do.

**Comment by mattludwigs** [comment]
I like that approach and make sense to me. Would there be a behavior for the consuming library to do what it needs to in order to report back that the hardware is okay?

**Comment by fhunleth** [comment]
I'm implementing it as a push rather than a pull. Anything that wants to report "ok" calls `VintageNet.PowerManager.PMControl.pet_watchdog(ifname)`.

**Comment by fhunleth** [comment]
I pushed the rewrite. The next step is to put in calls to pet the device watchdog in good places. For LTE, which is what's driving this, the good place looks like the signal strength/connection poller. In all of the LTE module failures I've seen so far, something happens with that code - like errors back from the AT commands. In the future, I think the network traffic monitor would be good for this since it will report whether anything has been sent and received on a network device without doing so itself.

**Comment by fhunleth** [comment]
I did some cleanup/rebasing. In the state that it's in right now, it will recover by resetting the LTE modem when I cause it to error indefinitely. `VintageNet.info` provides a one line status for what's happening with power management-enabled interfaces and how much time is left in each state. I think that it's finally getting close. There's some minor cleanup left and I'd like to break this PR up since I don't like my fix to ignore redundant network present notifications since it seems like that points to a bug.
